Other tests for a drift element

MMannieCorbisier opened the pull request:
Tests are added to check for low/high energies conservation along a drift for proton/electron pencil beam, to check for the conservation of the number of particles, of the rectilinear trajectory and for the match with the relativistic expression of the momentum.

MMannieCorbisier committed:
Other tests for a drift element

stewartboogert commented on 01_element/test_drift2.py:
```diff
@@ -0,0 +1,267 @@
+            ek_out=ek_out[mask]
+            ek_out=np.mean(ek_out)  
+
+        assert (pytest.approx(ek_out)==ek_in) & (pytest.approx(phi_out)==phi_in) & (pytest.approx(theta_out)==theta_in) & (n_in==n_out) & (pytest.approx(pz_out)==pz_rel)
```
Is there an actual test of the change of position, and/or test of the time?

stewartboogert commented on 01_element/test_drift2.py:
```diff
@@ -0,0 +1,267 @@
+            ek_out=ek_out[mask]
+            ek_out=np.mean(ek_out)  
+
+        assert (pytest.approx(ek_out)==ek_in) & (pytest.approx(phi_out)==phi_in) & (pytest.approx(theta_out)==theta_in) & (n_in==n_out) & (pytest.approx(pz_out)==pz_rel)
```
Same as above, position and time?

stewartboogert commented on 01_element/test_drift2.py:
```diff
@@ -0,0 +1,267 @@
+            a=f[name+'kineticEnergy'].array(library='np') # GeV
+            ek_out= np.array([x[0] for x in a if len(x)>0])
+            ek_out=ek_out[mask]
+            ek_out=np.mean(ek_out)  
```
Is there a way by writing a function to access sampler data using uproot. This will make reading the logic of this test much easier e.g. 

```
def getSamplerData(samplerName="d1", variable="phi", mask) :
    d =  [name+'kineticEnergy'].array(library='np') 
    d_out =  np.array([x[0] for x in d if x.size>0])
    d_out = d_out[mask]
    return d_out
```

stewartboogert reviewed (commented):
The test is good

1. it can be simplified a lot by adding a function to access the sampler data.
2. the assertion should include time and position

MMannieCorbisier committed:
adding of time and position tests

MMannieCorbisier commented:
Thanks for the review. 

I've added the requested changes:

- sampler access is now factorized in a helper function  
- tests now include time and position (via the length traveled and both angles)
- the overall test structure is simplified  

I also added the concept of the p-value as a possible generalization of the test. However, in this case, the errors come from numerical round-off rather than statistics, so no statistical test is performed in the function. Instead, it uses a simple “close enough” numerical comparison.

Let me know if you’d like me to adjust anything further.

MMannieCorbisier committed:
Correction of p-value test in test_drift2.py

MMannieCorbisier commented:
On second thought, I think that the test with the p-value function adds unnecessary complexity to the code, so I removed it for better clarity.

MMannieCorbisier committed:
Simplification of the code

stewartboogert commented on 01_element/test_drift2.py:
```diff
@@ -0,0 +1,111 @@
+            
+            t_out_mean=getSamplerData(f,samplerName,"T", mask)
+            
+            s_sampler_mean=getSamplerData(f,samplerName,"S", mask)
```
Remove blank lines

stewartboogert commented on 01_element/test_drift2.py:
```diff
@@ -0,0 +1,111 @@
+            a=f[samplerName+'parentID'].array(library='np')
+            parent_ID= np.array([x[0] for x in a if len(x)>0])
+            mask=parent_ID==0 #from initial beam only
+            n_out=np.sum(mask)
```
Still I feel a function can be used for this

stewartboogert commented on 01_element/test_drift2.py:
```diff
@@ -0,0 +1,111 @@
+            pz_frac= np.array([x[0] for x in a if len(x)>0])
+            pz_out=np.multiply(pz_frac,p_out)
+            pz_out=pz_out[mask]
+            pz_out_mean=np.mean(pz_out)
```
Same here about a function

MMannieCorbisier committed:
Add utility functions and remove blank lines

MMannieCorbisier committed:
Add utility functions and remove blank lines

MMannieCorbisier commented:
Comments addressed.
